File: scripts/apply_enrichment.py
#!/usr/bin/env python3
"""
H-Moon Hydro: Apply Enrichment Data
====================================
Applies brands and descriptions from deep_enrichment.json to WooCommerce products.

What gets applied:
  - Brands (421 products) -> pwb-brand taxonomy (ALWAYS applied)
  - Descriptions (48 products) -> Only where currently empty
  - Short descriptions (72 products) -> Only where currently empty

What gets SKIPPED:
  - Images (Shopify CDN URLs won't work on WooCommerce)
  - Prices (don't overwrite with competitor pricing)

Usage:
  python scripts/apply_enrichment.py           # Dry run (default)
  python scripts/apply_enrichment.py --confirm  # LIVE run
"""

import json, os, sys

BASE = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
CONFIRM = '--confirm' in sys.argv

SSH_HOST = os.getenv('HMOON_SSH_HOST')
SSH_USER = os.getenv('HMOON_SSH_USER')
SSH_PASS = os.getenv('HMOON_SSH_PASS')
SITE_DIR = os.getenv('HMOON_SITE_DIR', '~/hmoonhydro.com')


def run():
    import paramiko

    if not SSH_HOST or not SSH_USER or not SSH_PASS:
        raise SystemExit('Missing SSH env vars: HMOON_SSH_HOST, HMOON_SSH_USER, HMOON_SSH_PASS')

    dry_run = not CONFIRM
    mode = 'DRY RUN' if dry_run else 'LIVE'

    # Load enrichment data
    enrichment_path = os.path.join(BASE, 'outputs', 'deep_enrichment.json')
    with open(enrichment_path, 'r', encoding='utf-8') as f:
        data = json.load(f)

    print('=== Apply Enrichment: ' + mode + ' ===')
    print('Total products in enrichment: ' + str(len(data)))

    # Build the enrichment payload - only what we want to apply
    payload = {}
    for pid, prod in data.items():
        entry = {}
        if prod.get('brand'):
            entry['brand'] = prod['brand']
        if prod.get('description'):
            entry['description'] = prod['description']
        if prod.get('short_description'):
            entry['short_description'] = prod['short_description']
        if entry:
            payload[pid] = entry

    brand_count = sum(1 for v in payload.values() if v.get('brand'))
    desc_count = sum(1 for v in payload.values() if v.get('description'))
    short_count = sum(1 for v in payload.values() if v.get('short_description'))

    print('Brands to set: ' + str(brand_count))
    print('Descriptions (fill empty): ' + str(desc_count))
    print('Short descriptions (fill empty): ' + str(short_count))
    print()

    # Build PHP
    dry_val = 'true' if dry_run else 'false'
    
    p = []
    p.append('<?php')
    p.append('wp_set_current_user(1);')
    p.append('$dry = ' + dry_val + ';')
    p.append('$data = json_decode(file_get_contents(__DIR__ . "/enrichment_payload.json"), true);')
    p.append('$stats = ["brands" => 0, "desc" => 0, "short_desc" => 0, "skipped" => 0, "not_found" => 0];')
    p.append('')
    p.append('echo "Processing " . count($data) . " products...\\n";')
    p.append('')
    p.append('foreach ($data as $pid => $fields) {')
    p.append('    $post = get_post($pid);')
    p.append('    if (!$post || $post->post_type !== "product") {')
    p.append('        $stats["not_found"]++;')
    p.append('        continue;')
    p.append('    }')
    p.append('')
    p.append('    // Set brand via pwb-brand taxonomy')
    p.append('    if (!empty($fields["brand"])) {')
    p.append('        $brand = $fields["brand"];')
    p.append('        if ($dry) {')
    p.append('            echo "  [DRY] $pid: set brand \\"$brand\\"\\n";')
    p.append('        } else {')
    p.append('            // Check if brand term exists, create if needed')
    p.append('            $bt = get_term_by("name", $brand, "pwb-brand");')
    p.append('            if (!$bt) {')
    p.append('                $r = wp_insert_term($brand, "pwb-brand");')
    p.append('                if (is_wp_error($r)) {')
    p.append('                    echo "  WARN: Could not create brand \\"$brand\\": " . $r->get_error_message() . "\\n";')
    p.append('                }')
    p.append('            }')
    p.append('            wp_set_object_terms($pid, $brand, "pwb-brand", false);')
    p.append('            $stats["brands"]++;')
    p.append('        }')
    p.append('    }')
    p.append('')
    p.append('    // Set description only if currently empty')
    p.append('    if (!empty($fields["description"])) {')
    p.append('        $current = $post->post_content;')
    p.append('        if (empty(trim($current))) {')
    p.append('            if ($dry) {')
    p.append('                echo "  [DRY] $pid: set description (" . strlen($fields["description"]) . " chars)\\n";')
    p.append('            } else {')
    p.append('                wp_update_post(array("ID" => $pid, "post_content" => $fields["description"]));')
    p.append('                $stats["desc"]++;')
    p.append('            }')
    p.append('        } else {')
    p.append('            $stats["skipped"]++;')
    p.append('        }')
    p.append('    }')
    p.append('')
    p.append('    // Set short description only if currently empty')
    p.append('    if (!empty($fields["short_description"])) {')
    p.append('        $current_short = $post->post_excerpt;')
    p.append('        if (empty(trim($current_short))) {')
    p.append('            if ($dry) {')
    p.append('                echo "  [DRY] $pid: set short_description (" . strlen($fields["short_description"]) . " chars)\\n";')
    p.append('            } else {')
    p.append('                wp_update_post(array("ID" => $pid, "post_excerpt" => $fields["short_description"]));')
    p.append('                $stats["short_desc"]++;')
    p.append('            }')
    p.append('        } else {')
    p.append('            $stats["skipped"]++;')
    p.append('        }')
    p.append('    }')
    p.append('}')
    p.append('')
    p.append('echo "\\n=== SUMMARY ===\\n";')
    p.append('$mode_str = $dry ? "DRY RUN" : "LIVE";')
    p.append('echo "Mode: $mode_str\\n";')
    p.append('echo "Brands set: " . $stats["brands"] . "\\n";')
    p.append('echo "Descriptions set: " . $stats["desc"] . "\\n";')
    p.append('echo "Short descriptions set: " . $stats["short_desc"] . "\\n";')
    p.append('echo "Skipped (already had content): " . $stats["skipped"] . "\\n";')
    p.append('echo "Not found: " . $stats["not_found"] . "\\n";')
    p.append('echo "Done.\\n";')

    php_code = '\n'.join(p)

    # Connect
    ssh = paramiko.SSHClient()
    ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
    ssh.connect(SSH_HOST, username=SSH_USER, password=SSH_PASS)

    remote_dir = SITE_DIR + '/wp-content'
    remote_php = remote_dir + '/apply_enrichment.php'
    remote_json = remote_dir + '/enrichment_payload.json'

    # Upload payload JSON + PHP
    sftp = ssh.open_sftp()
    payload_str = json.dumps(payload, ensure_ascii=False)
    with sftp.open(remote_json, 'w') as f:
        f.write(payload_str)
    with sftp.open(remote_php, 'w') as f:
        f.write(php_code)
    sftp.close()

    print('Uploaded PHP + JSON payload (' + str(len(payload_str) // 1024) + ' KB)')

    # Execute
    print('Executing (' + mode + ')...\n')
    cmd = 'cd ' + SITE_DIR + ' && wp eval-file wp-content/apply_enrichment.php'
    stdin, stdout, stderr = ssh.exec_command(cmd, timeout=600)

    output = stdout.read().decode('utf-8', errors='replace')
    errors = stderr.read().decode('utf-8', errors='replace')

    # Print (truncated for dry run since lots of lines)
    lines = output.split('\n')
    if len(lines) > 60 and dry_run:
        for line in lines[:30]:
            print(line.encode('ascii', errors='replace').decode('ascii'))
        print('... (' + str(len(lines) - 60) + ' more lines) ...')
        for line in lines[-30:]:
            print(line.encode('ascii', errors='replace').decode('ascii'))
    else:
        for line in lines:
            print(line.encode('ascii', errors='replace').decode('ascii'))

    if errors:
        for line in errors.strip().split('\n'):
            if 'Warning' not in line and 'Notice' not in line and 'Deprecated' not in line:
                print('STDERR: ' + line.encode('ascii', errors='replace').decode('ascii'))

    # Cleanup
    ssh.exec_command('rm ' + remote_php + ' ' + remote_json)
    ssh.close()

    # Save output
    slug = mode.lower().replace(' ', '_')
    out_path = os.path.join(BASE, 'outputs', 'enrichment_' + slug + '_output.txt')
    with open(out_path, 'w', encoding='utf-8') as f:
        f.write(output)
    print('\nOutput saved: ' + out_path)


if __name__ == '__main__':
    run()

File: scripts/apply_enrichment.php
<?php
/**
 * apply_enrichment.php — Apply scraped enrichment data to WooCommerce products
 * Reads enrichment_to_apply.json uploaded alongside this script
 * 
 * Handles: image_url, gallery_urls, brand, short_description
 */
wp_set_current_user(1);
require_once ABSPATH . 'wp-admin/includes/media.php';
require_once ABSPATH . 'wp-admin/includes/file.php';
require_once ABSPATH . 'wp-admin/includes/image.php';

$dry_run = getenv('CONFIRM') !== '1';
echo $dry_run ? "=== DRY RUN ===\n\n" : "=== LIVE MODE ===\n\n";

// Read enrichment data
$json_file = dirname(__FILE__) . '/enrichment_to_apply.json';
if (!file_exists($json_file)) {
    echo "ERROR: enrichment_to_apply.json not found at: {$json_file}\n";
    exit(1);
}
$enrichments = json_decode(file_get_contents($json_file), true);
echo "Enrichment records: " . count($enrichments) . "\n\n";

$stats = array('image'=>0, 'gallery'=>0, 'brand'=>0, 'short_desc'=>0, 'errors'=>0);

foreach ($enrichments as $i => $en) {
    $pid = $en['id'];
    $title = $en['title'];
    $apply = $en['apply'];
    $score = $en['score'];
    
    // Only auto-apply high-confidence matches (>= 0.80)
    if ($score < 0.60) continue;
    
    $changes = array();
    
    // --- MAIN IMAGE ---
    if (!empty($apply['image_url'])) {
        $url = $apply['image_url'];
        $changes[] = "image: " . basename(parse_url($url, PHP_URL_PATH));
        
        if (!$dry_run) {
            $tmp = download_url($url, 30);
            if (!is_wp_error($tmp)) {
                $file_array = array(
                    'name' => sanitize_file_name(basename(parse_url($url, PHP_URL_PATH))),
                    'tmp_name' => $tmp,
                );
                $att_id = media_handle_sideload($file_array, $pid, $title);
                if (!is_wp_error($att_id)) {
                    set_post_thumbnail($pid, $att_id);
                    $stats['image']++;
                } else {
                    echo "  ERROR image #{$pid}: " . $att_id->get_error_message() . "\n";
                    $stats['errors']++;
                }
            } else {
                echo "  ERROR download #{$pid}: " . $tmp->get_error_message() . "\n";
                $stats['errors']++;
                @unlink($tmp);
            }
        } else {
            $stats['image']++;
        }
    }
    
    // --- GALLERY IMAGES ---
    if (!empty($apply['gallery_urls'])) {
        $gallery_ids = array();
        $existing = get_post_meta($pid, '_product_image_gallery', true);
        if (!empty($existing)) {
            $gallery_ids = explode(',', $existing);
        }
        
        $count = count($apply['gallery_urls']);
        $changes[] = "gallery: +{$count} images";
        
        if (!$dry_run) {
            $added = 0;
            foreach ($apply['gallery_urls'] as $gurl) {
                if ($added >= 5) break; // Max 5 gallery images per product
                
                $tmp = download_url($gurl, 30);
                if (!is_wp_error($tmp)) {
                    $file_array = array(
                        'name' => sanitize_file_name(basename(parse_url($gurl, PHP_URL_PATH))),
                        'tmp_name' => $tmp,
                    );
                    $att_id = media_handle_sideload($file_array, $pid);
                    if (!is_wp_error($att_id)) {
                        $gallery_ids[] = $att_id;
                        $added++;
                    }
                } else {
                    @unlink($tmp);
                }
            }
            if (!empty($gallery_ids)) {
                update_post_meta($pid, '_product_image_gallery', implode(',', array_unique($gallery_ids)));
                $stats['gallery']++;
            }
        } else {
            $stats['gallery']++;
        }
    }
    
    // --- BRAND ---
    if (!empty($apply['brand'])) {
        $brand_name = $apply['brand'];
        $changes[] = "brand: {$brand_name}";
        
        if (!$dry_run) {
            $bt = get_term_by('name', $brand_name, 'pwb-brand');
            if (!$bt) {
                $r = wp_insert_term($brand_name, 'pwb-brand');
                if (!is_wp_error($r)) {
                    $bt = get_term_by('term_id', $r['term_id'], 'pwb-brand');
                }
            }
            if ($bt) {
                wp_set_object_terms($pid, array($bt->term_id), 'pwb-brand', false);
                $stats['brand']++;
            }
        } else {
            $stats['brand']++;
        }
    }
    
    // --- SHORT DESCRIPTION ---
    if (!empty($apply['short_description'])) {
        $short = $apply['short_description'];
        $changes[] = "short_desc: " . strlen($short) . " chars";
        
        if (!$dry_run) {
            wp_update_post(array('ID' => $pid, 'post_excerpt' => $short));
            $stats['short_desc']++;
        } else {
            $stats['short_desc']++;
        }
    }
    
    if (!empty($changes)) {
        $label = ($score >= 0.80) ? "HIGH" : "MED";
        echo "  #{$pid} [{$label} {$score}] " . implode(' | ', $changes) . "\n";
    }
    
    // Progress update every 50
    if (($i + 1) % 50 === 0) {
        echo "  ... processed " . ($i + 1) . "/" . count($enrichments) . "\n";
    }
}

echo "\n=== RESULTS ===\n";
echo "  Images set:         {$stats['image']}\n";
echo "  Galleries added:    {$stats['gallery']}\n";
echo "  Brands assigned:    {$stats['brand']}\n";
echo "  Short descs added:  {$stats['short_desc']}\n";
echo "  Errors:             {$stats['errors']}\n";
echo "\n";
if ($dry_run) echo "*** DRY RUN — use CONFIRM=1 to apply ***\n";
